source/extensions/filters/network/redis_proxy/config.cc

```diff
@@ -48,24 +60,42 @@ Network::FilterFactoryCb RedisProxyFilterConfigFactory::createFilterFactoryFromP
       std::make_shared<ProxyFilterConfig>(proto_config, context.scope(), context.drainDecision(),
                                           context.runtime(), context.api(), cache_manager_factory);
 
-  envoy::extensions::filters::network::redis_proxy::v3::RedisProxy::PrefixRoutes prefix_routes(
-      proto_config.prefix_routes());
-
-  // Set the catch-all route from the settings parameters.
-  if (prefix_routes.routes_size() == 0 && !prefix_routes.has_catch_all_route()) {
-    throw EnvoyException("cannot configure a redis-proxy without any upstream");
-  }
-
   absl::flat_hash_set<std::string> unique_clusters;
-  for (auto& route : prefix_routes.routes()) {
-    addUniqueClusters(unique_clusters, route);
+
+  envoy::extensions::filters::network::redis_proxy::v3::RedisProxy::PrefixRoutes prefix_routes;
+  envoy::extensions::filters::network::redis_proxy::v3::RedisProxy::HashSlotRoutes hash_slot_routes;
+
+  // Add upstream clusters
+  if (proto_config.has_prefix_routes()) {
+    prefix_routes = proto_config.prefix_routes();
+    // Set the catch-all route from the settings parameters.
+    if (prefix_routes.routes_size() == 0 && !prefix_routes.has_catch_all_route()) {
+      throw EnvoyException("cannot configure a redis-proxy without any upstream");
+    }
+
+    for (auto& route : prefix_routes.routes()) {
+      addUniqueClustersPrefixRoutes(unique_clusters, route);
+    }
+    addUniqueClustersPrefixRoutes(unique_clusters, prefix_routes.catch_all_route());
+  } else if (proto_config.has_hash_slot_routes()) {
+    hash_slot_routes = proto_config.hash_slot_routes();
+    // Set the catch-all route from the settings parameters.
+    if (hash_slot_routes.routes_size() == 0 && !hash_slot_routes.has_catch_all_route()) {
+      throw EnvoyException("cannot configure a redis-proxy without any upstream");
+    }
+
+    for (auto& route : hash_slot_routes.routes()) {
+      addUniqueClustersHashSlotRoutes(unique_clusters, route);
+    }
+    addUniqueClustersHashSlotRoutes(unique_clusters, hash_slot_routes.catch_all_route());
+  } else {
+    throw EnvoyException("please define a router for upstream clusters");
```

Comment:
Oh you are right, nice catch!

Reply:
@mattklein123 Checked again, I think prefix routing is required before as upstream redis cluster needs to be defined inside prefix routing. Previous error msg is "cannot configure a redis-proxy without any upstream" and right now it will become this. We can change the error msg to match that but please suggest.